crates/ruff/src/message/mod.rs

```diff
@@ -20,49 +22,64 @@ pub use grouped::GroupedEmitter;
 pub use json::JsonEmitter;
 pub use junit::JunitEmitter;
 pub use pylint::PylintEmitter;
-pub use rustpython_parser::ast::Location;
 pub use text::TextEmitter;
 
 use crate::jupyter::JupyterIndex;
+use crate::registry::AsRule;
 use ruff_diagnostics::{Diagnostic, DiagnosticKind, Fix};
-use ruff_python_ast::source_code::SourceFile;
+use ruff_python_ast::source_code::{SourceFile, SourceLocation};
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct Message {
     pub kind: DiagnosticKind,
-    pub location: Location,
-    pub end_location: Location,
+    pub range: TextRange,
     pub fix: Fix,
     pub file: SourceFile,
-    pub noqa_row: usize,
+    pub noqa_offset: TextSize,
 }
 
 impl Message {
-    pub fn from_diagnostic(diagnostic: Diagnostic, file: SourceFile, noqa_row: usize) -> Self {
+    pub fn from_diagnostic(
+        diagnostic: Diagnostic,
+        file: SourceFile,
+        noqa_offset: TextSize,
+    ) -> Self {
         Self {
+            range: diagnostic.range(),
             kind: diagnostic.kind,
-            location: Location::new(diagnostic.location.row(), diagnostic.location.column() + 1),
-            end_location: Location::new(
-                diagnostic.end_location.row(),
-                diagnostic.end_location.column() + 1,
-            ),
             fix: diagnostic.fix,
             file,
-            noqa_row,
+            noqa_offset,
         }
     }
 
     pub fn filename(&self) -> &str {
         self.file.name()
     }
+
+    pub fn compute_start_location(&self) -> SourceLocation {
```

Comment:
It can be, that's why I called it `compute` to motivate callers to hold on to it, rather than calling `start_location` over and over again. It's somewhat expensive because:

* It needs to compute the `LineIndex` if it the `index` hasn't yet been computed for this file.
* It performs a binary search to find the index fo the given offset. This isn't expensive, but it's still better to hold on to the location rather than performing the binary search over and over again :)